engine/crates/composition/src/ingest_subgraph.rs

```diff
@@ -49,13 +53,33 @@ fn ingest_top_level_definitions(
                 let directives = subgraphs.new_directive_site();
 
                 let definition_id = match &type_definition.node.kind {
-                    ast::TypeKind::Object(_) => subgraphs.push_definition(
-                        subgraph_id,
-                        type_name,
-                        DefinitionKind::Object,
-                        description,
-                        directives,
-                    ),
+                    ast::TypeKind::Object(_) => {
+                        let definition_id = subgraphs.push_definition(
+                            subgraph_id,
+                            type_name,
+                            DefinitionKind::Object,
+                            description,
+                            directives,
+                        );
+
+                        match root_type_matcher.match_name(type_name) {
+                            RootTypeMatch::Query => {
+                                subgraphs.set_query_type(subgraph_id, definition_id);
+                            }
+                            RootTypeMatch::Mutation => {
+                                subgraphs.set_mutation_type(subgraph_id, definition_id);
+                            }
+                            RootTypeMatch::Subscription => {
+                                subgraphs.set_subscription_type(subgraph_id, definition_id);
+                            }
+                            RootTypeMatch::NotRootButHasDefaultRootName => {
+                                subgraphs.push_ingestion_diagnostic(subgraph_id, format!("The {type_name} type has the default name for a root but is itself not a root. This is not valid in a federation context."));
```

Comment:
Just curious: did we define this rule or is it standard elsewhere?

Reply:
It's a rule replicated from the implementations we compare with. The message is not exactly the same but it's the same rule.